### PR TITLE
feat: Generate index metadata.

### PR DESCRIPTION
**Pull request by zsol**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #118
* #117

This is going to be useful for the frontend: it can load `/index/.metadata` instead of crawling through all files at build time to power:
- listing all projects (to be removed later)
- search projects by name (to be added)
- show latest indexed projects (to be added)
- show most popular projects (to be added)

### CODE REVIEW

**Comment by zsol** [comment]
@abesto when you have time, wdyt of handling this on the frontend side?

TLDR - the indexer generates a `.metadata` json file which should contain everything necessary to render the "landing page" without having to iterate over all the individual `${project}.json` files.

We could process `.metadata` at compile time like we do today, or we can just send a request at runtime. Or maybe both? Is that premature optimization? :)

**Comment by abesto** [comment]
Ah nice. FWIW, the index page is statically generated (and is currently the only such page), so this would (only) speed up build times. Are you experiencing a long(ish) www build time? Either way, doing this should be fairly straightforward. I don't see a way to get do "shared" stacks with `ghstack` (educate me?), so I guess I can do it after this lands?

**Comment by zsol** [comment]
Hmm, thanks for making me think :D yeah none of this is really needed atm. build time is still ~2 seconds with ~550 projects.

File: py_wtf/cli/__init__.py [hunk 1]
```diff
@@ -145,25 +150,29 @@ async def generate_test_index(dir: str | None) -> None:
...
         proj = Project(
-            ProjectName(proj_dir.name),
+            proj_name,
             metadata=proj_info,
             modules=mods,
             documentation=[Documentation(proj_info.summary or "")],
         )
         (out_dir / f"{proj.name}.json").write_text(converter.dumps(_sort(proj)))
+    repo = ProjectRepository(out_dir)
+    repo.write_index(1662219954)
```
Comment (abesto) [inline py_wtf/cli/__init__.py:177]: `time.time()`? :D

**Review by abesto — APPROVED** [review]
Hm. AFAICT this traverses the existing dump, yes? That's surprising, I was expecting generating the metadata file to be a side-effect of the indexing process using the data already in memory. But also, pipeline-ing is good. <3 for sorting by latest updated / top (by dependent count).

File: py_wtf/cli/__init__.py (hunk 1 above)
Comment (zsol) [reply py_wtf/cli/__init__.py:177]: no it has to be deterministic :)

**Comment by zsol** [comment]
It uses data already in memory just to avoid having to `loads(foo.read_text())`. After an indexing run there might be projects not loaded into memory though, and we'd want to consider those too for the metadata.

Since indexing _always_ writes the json to disk, the easiest is to iterate over all json files on disk to build the inventory.

File: py_wtf/cli/__init__.py (hunk 1 above)
Comment (abesto) [reply py_wtf/cli/__init__.py:177]: Not obvious to me on 15 seconds of thinking why. Why? And also, why is a hardcoded timestamp the right (or even a good) "hash"?

Comment (zsol) [reply py_wtf/cli/__init__.py:177]: Because there's a pre-commit hook and CI check to make sure the test index is always up-to-date

**Pull request merged**